test/integration/analysis.js

```diff
@@ -120,6 +123,52 @@ describe('workflow', function() {
             });
         });
 
+        it('PRECHECK query should be inside a read-only transaction', function(done) {
+            var TEST_SOURCE_TYPE = 'test-source';
+            var TestSource = Node.create(TEST_SOURCE_TYPE, {
+                table: Node.PARAM.STRING()
+            }, {cache: true});
+            TestSource.prototype.sql = function() {
+                return 'select * from ' + this.table;
+            };
+            TestSource.prototype.preCheckQuery = function() {
+                return 'INSERT INTO ' + this.table + ' VALUES (999, NULL, NULL, 100)';
+            };
+
+            var supportsTypeFn = Factory.prototype._supportsType;
+            var getNodeClassFn = Factory.prototype.getNodeClass;
+
+            Factory.prototype._supportsType = function() {
+                return true;
+            };
+
+            Factory.prototype.getNodeClass = function() {
+                return TestSource;
+            };
+
+            var definition = {
+                type: TEST_SOURCE_TYPE,
+                params: {
+                    table: 'airbnb_rooms'
+                }
+            };
+            testConfig.batch.inlineExecution = true;
+            Analysis.create(testConfig, definition, function(err, analysis) {
+                Factory.prototype._supportsType = supportsTypeFn;
+                Factory.prototype.getNodeClass = getNodeClassFn;
+                assert.ifError(err);
+                var catalogQuery = ['SELECT last_error_message ',
```

Comment:
😈  This is a bit subtle. The first time the analysis is executed it doesn't know anything about the error because, as your test reflects, the error message is picked up from the analysis catalog. Even if the queries are executed inline, the behaviour is still async, and until the second attempt to create the analysis, the error message won't be retrieved from the catalog table.

In order to work around that situation, the test helper provides you with a `createAnalyses` method that instantiates the analysis twice in order to complete that double cycle, making it possible to pickup the error message from the catalog table.

For this test, instead of using the `Analysis.create` method from the library, I recommend you to use the test helper, as it will assist you by doing the two analysis instantiations required to complete the cycle, and it will populate the error into the node.

```diff
diff --git a/test/integration/analysis.js b/test/integration/analysis.js
index 460a222..f21d054 100644
--- a/test/integration/analysis.js
+++ b/test/integration/analysis.js
@@ -152,21 +152,16 @@ describe('workflow', function() {
                     table: 'airbnb_rooms'
                 }
             };
-            testConfig.batch.inlineExecution = true;
-            Analysis.create(testConfig, definition, function(err, analysis) {
+            var config = testConfig.create({ batch: { inlineExecution: true } });
+            testHelper.createAnalyses(definition, config, function(err, analysis) {
                 Factory.prototype._supportsType = supportsTypeFn;
                 Factory.prototype.getNodeClass = getNodeClassFn;
                 assert.ifError(err);
-                var catalogQuery = ['SELECT last_error_message ',
-                                    'FROM cdb_analysis_catalog ',
-                                    'WHERE node_id = \'' + analysis.id() + '\''].join('\n');
-                testHelper.executeQuery(catalogQuery, function(err, results) {
-                    assert.ifError(err);
-                    assert.equal(1,results.rows.length);
-                    assert.equal('cannot execute INSERT in a read-only transaction',
-                                 results.rows[0].last_error_message);
-                    done();
-                });
+                assert.equal(
+                    analysis.getRoot().getErrorMessage(),
+                    'cannot execute INSERT in a read-only transaction'
+                );
+                done();
             });
         });
```